pallets/pallet-bonded-coins/src/lib.rs

```diff
@@ -230,5 +313,66 @@ pub mod pallet {
 		}
 	}
 
-	impl<T: Config> Pallet<T> {}
+	impl<T: Config> Pallet<T>
+	where
+		<CurveParameterTypeOf<T> as Fixed>::Bits: Copy + ToFixed + AddAssign + BitOrAssign + ShlAssign,
+	{
+		fn calculate_collateral(
+			low: CurveParameterTypeOf<T>,
+			high: CurveParameterTypeOf<T>,
+			passive_supply: PassiveSupply<CurveParameterTypeOf<T>>,
+			curve: &Curve<CurveParameterTypeOf<T>>,
+		) -> Result<CollateralCurrencyBalanceOf<T>, ArithmeticError> {
+			let normalized_costs = curve.calculate_costs(low, high, passive_supply)?;
+
+			let collateral_denomination = 10u128
+				.checked_pow(T::CollateralCurrency::decimals(T::CollateralAssetId::get()).into())
+				.ok_or(ArithmeticError::Overflow)?;
+
+			let real_costs = normalized_costs
+				.checked_mul(CurveParameterTypeOf::<T>::from_num(collateral_denomination))
+				.ok_or(ArithmeticError::Overflow)?
+				// should never fail
+				.checked_to_num::<u128>()
+				.ok_or(ArithmeticError::Overflow)?
+				.saturated_into();
+
+			Ok(real_costs)
+		}
+
+		fn calculate_normalized_passive_issuance(
+			bonded_currencies: &[FungiblesAssetIdOf<T>],
+			denomination: u8,
+			currency_idx: usize,
+		) -> Result<(CurveParameterTypeOf<T>, PassiveSupply<CurveParameterTypeOf<T>>), DispatchError> {
+			let currencies_total_supply = bonded_currencies
+				.iter()
+				.map(|currency_id| Self::get_fungible_supply::<T::AssetId, T::Fungibles>(currency_id.to_owned()))
+				.collect::<Vec<_>>();
+
+			let normalized_total_issuances = currencies_total_supply
```

Comment:
The more I think about this problem, the more I prefer assuming the worst case. Any unused weight would be refunded to the user. During benchmarking, I would also see how many currencies I can allow. 

Any concerns?

Reply:
It probably does not apply to our runtime specifically, but the main problem there is that even if you know there is, for instance, a single currency bonded, but the worst case is 1000, the call could fail to go through if the worst case is too large, even if in practice it would go through. That was the main problem with the staking pallet, and in general I tend to prefer a solution where the weight calculated is as accurate as possible.

Reply:
I see. I have added the benchmark parameter in: [90b1572](https://github.com/KILTprotocol/kilt-node/pull/770/commits/90b15725cfad44ca3e460bf82e9f59dab34e4599) 

the benchmark logic would be similar to the `set_inflation` tx in the staking pallet

Reply:
I like a combination of both: require callers to provide a currency number that is at least the number of actual currencies and refund some part of the fees if the actual number is less than that. This even allows to opt for worst case weight calculations on the caller side by just passing in the MaxCurrencies value everywhere.